File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from collections import deque
from sys import stdin as s

n, m, v = map(int, s.readline().strip().split())
graph = [list() for i in range(n + 1)]
visited = [False] * (n + 1)
for i in range(m):
    a, b = map(int, s.readline().strip().split())
    if b not in graph[a]:
        graph[a].append(b)
    if a not in graph[b]:
        graph[b].append(a)
for i in graph:
    i.sort()

def dfs(graph, start):
    visited = [False] * (n + 1)
    lst = deque([start])

    while lst:
        node = lst.pop()
        if not visited[node]:
            print(node, end=" ")
            visited[node] = True
            lst.extend(graph[node][::-1])


def bfs(graph, start):
    visited = [False] * (n + 1)
    lst = deque([start])
    # 초기값 설정.
    while lst:
        node = lst.popleft()
        if not visited[node]:
            print(node, end=" ")
        visited[node] = True
        for i in graph[node]:
            if not visited[i]:

                lst.append(i)

dfs(graph,v)
print()
bfs(graph,v)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.*;
import java.util.*;

public class Main {

    static boolean[] visited;
    static ArrayList<Integer>[] graph;

    public static void main(String[] args) throws IOException {
        // 입력을 받기 위해 BufferedReader 사용
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // 첫 줄에서 n, m, v를 입력 받음
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int v = Integer.parseInt(st.nextToken());

        // 그래프 초기화
        graph = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }

        // 간선 정보를 입력 받아 그래프 구성
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            if (!graph[a].contains(b)) {
                graph[a].add(b);
            }
            if (!graph[b].contains(a)) {
                graph[b].add(a);
            }
        }

        // 각 노드의 인접 리스트를 정렬
        for (int i = 1; i <= n; i++) {
            Collections.sort(graph[i]);
        }

        // DFS 수행
        visited = new boolean[n + 1];
        dfs(v);
        System.out.println();

        // BFS 수행
        visited = new boolean[n + 1];
        bfs(v);
    }

    // DFS 메서드
    public static void dfs(int node) {
        visited[node] = true;
        System.out.print(node + " ");
        for (int next : graph[node]) {
            if (!visited[next]) {
                dfs(next);
            }
        }
    }

    // BFS 메서드
    public static void bfs(int start) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(start);
        visited[start] = true;

        while (!queue.isEmpty()) {
            int node = queue.poll();
            System.out.print(node + " ");
            for (int next : graph[node]) {
                if (!visited[next]) {
                    visited[next] = true;
                    queue.add(next);
                }
            }
        }
    }
}
